### PR TITLE
[MRG] Added overloads to Colour for to support values between 0 - 255

### PR DESCRIPTION
**Pull request by gurrenm3**
Added the following:
• 1 constructor overload to support integer RGB values between 0 - 255
• 2 implicit operators to support converting between System.Color and NoManSky's Colour class
• XML Doc for everything in the class

### CODE REVIEW

File: libMBIN/Source/NMS/BaseTypes/Colour.cs [hunk 1]
```diff
@@ -1,33 +1,104 @@
...
+            this.R = R / 255;
+            this.G = G / 255;
+            this.B = B / 255;
+            this.A = A;
+        }
+
         public Colour() { }
+
+        /// <summary>
+        /// Returns a formatted string for this Color.
```
Comment (monkeyman192) [inline libMBIN/Source/NMS/BaseTypes/Colour.cs]: Might be good to mention that the format will be NMS-style colour, (ie. floats for all values)

Comment (gurrenm3) [reply libMBIN/Source/NMS/BaseTypes/Colour.cs]: Done

**Comment by cmkushnir** [comment]
1) It is non-standard to mix types for color channels i.e. rgb as int and alpha as float.  Should make alpha param an int as well in constructor.
2) There are mult color types in .NET, the two most common are System.Drawing.Color and System.Windows.Media.Color, they are distinct types and if you're going to support one you should support both.  Most of the WPF stuff uses System.Windows.Media.Color, most of the Forms stuff uses System.Drawing.Color e.g. NMSMB uses WPF and System.Windows.Media.Color, but yours and other may still be Forms based.
See: [https://stackoverflow.com/questions/2428930/what-is-the-difference-between-system-drawing-color-and-system-windows-media-col](https://stackoverflow.com/questions/2428930/what-is-the-difference-between-system-drawing-color-and-system-windows-media-col)
3) Since there are mult .NET color classes you shoud fully qualify the names in the code to make it clear which is being used at a given point.
4) ToString should follow how System.Windows.Media.Color and|or System.Drawing.Color ToString work, not use a new format.
See: [https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color.tostring?view=net-6.0](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color.tostring?view=net-6.0)
[https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.color.tostring?view=windowsdesktop-6.0](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.color.tostring?view=windowsdesktop-6.0)
5) Noticed that there are only a couple places still using NMSDescription, although doesn't really belong in this pr I mention because this pr makes an effort to remove them from libMBIN Colour, they should be removed from the remaining code as well i.e. EmptyNode, GcProjectileData.BehaviourFlags, TkMetadataReadMask.MetadataReadMask.

**Comment by cmkushnir** [comment]
6) Should provide explit methods to convert to|from the 2 .NET color types, for cases where implicit conversion may make the code unclear.

**Comment by monkeyman192** [comment]
WRT point 5, sorry, that was just an ask from me because I figured if xml summaries were being added, might as well just remove the NMSDescription here. I'm half tempted to just remove them across the board since it never really took off and was a bit of a failed idea.

**Comment by cmkushnir** [comment]
<strike>The problem w/ adding the color conversions is that one is forms specific and one is wpf specific.  that means libmbin now becomes tied to both frameworks (?), and i think this might be the first code that makes that requirement, as opposed to keeping it only dep on core.</strike>
... [https://devblogs.microsoft.com/dotnet/windows-forms-designer-for-net-core-released/](https://devblogs.microsoft.com/dotnet/windows-forms-designer-for-net-core-released/) ... i did not know this, thought forms was still specific to framework only.

**Comment by gurrenm3** [comment]
I've taken care of all the things you mentioned. I was unable to support System.Windows.Media.Color because I would need to add an assembly reference

**Comment by cmkushnir** [comment]
ToString is still very non-standard, it's coded for a specific use, not a generic one.  The current ToString will not produce what an app would want to use in most cases. Easiest solution is to have ToString convert libmbin Colour to a System.Drawing.Color then return it's ToString value.  If you need a different format then it should be considered impl specific and you should write an extension method in your code.

File: libMBIN/Source/NMS/BaseTypes/Colour.cs [hunk 2]
```diff
@@ -11,23 +13,99 @@ namespace libMBIN.NMS
...
+        /// <summary>
+        /// The Alpha component of the colour.
+        /// </summary>
         [NMSDescription("The Alpha component of the colour")]
         public float A;
 
-        public Colour(float R, float G, float B, float A)
+
+        /// <summary>
+        /// Creates a Color while providing it with float RGB values.
```
Comment (monkeyman192) [inline libMBIN/Source/NMS/BaseTypes/Colour.cs]: `Colour` (I think there are a few other places with the American spelling also)

**Comment by monkeyman192** [comment]
> 1. It is non-standard to mix types for color channels i.e. rgb as int and alpha as float.  Should make alpha param an int as well in constructor.

I was going to disagree with this, and then after reading more I realise that it's the right move having all as ints...
In js/css world, it would be RGB as int, alpha as float as standard.
In .net world it seems that all 4 as ints is standard.

Looks like the system color uses all 4 ints, so yeah, makes sense

**Comment by gurrenm3** [comment]
> ToString is still very non-standard, it's coded for a specific use, not a generic one. The current ToString will not produce what an app would want to use in most cases. Easiest solution is to have ToString convert libmbin Colour to a System.Drawing.Color then return it's ToString value. If you need a different format then it should be considered impl specific and you should write an extension method in your code.

I did ToString on a System.Drawing.Color and this is how it printed, aside from being in float form. I've updated it so it's printing exactly like System.Drawing.Color.ToString() and using whole numbers between 0 - 255. 
I'd prefer to avoid instantiating a Color just to use ToString, so instead I'm multiplying by hand

**Comment by cmkushnir** [comment]
Ok, my bad, never seen a color formatted that way, just assumed it was custom (been a while since used forms).  If the System.Drawing.Color is that bad then may as well do what you want as nobody would use something in that form.  I expected it to be either the name of a well known color, or in hex form, or support a format string which would be ideal.

**Comment by gurrenm3** [comment]
I'm a bit unsure what you're looking for. Can you give me an example?
It doesn't matter too much to me what it looks like, as long as ToString clearly lets you know what the 4 values are. I also imagine that printing the values as floats would be more helpful, since the Colour class uses float values.
It might also be good to indicate that it is a color, since it would be difficult to distinguish between this and Vector4/Quaternion

**Comment by cmkushnir** [comment]
Saying it would have been nice to use a standard format but if it's so bad
that it will never be used then do what you want.

**Comment by monkeyman192** [comment]
Looking at the ToString doc is seems like it will return the name of the colour if it was instantiated from one, otherwise it will return it in the form that is in this PR
cf. https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color.tostring?view=net-6.0#returns
Edit: I think if we do want to use this same format, then maybe it makes more sense to return the float value since that's what the "natural" representation is in NMS

**Comment by gurrenm3** [comment]
I talked with cmk and we came up with a solution that allows people to specify their own format.

**Comment by cmkushnir** [comment]
Instead of the constructor clamping param, just accept what is passed i.e. don't do this in cons:
this.R = MathHelper.Clampf(R, 0, 1);  // just use this.R = R;
It can be useful to be able to hold out-of-bound data from an intermediate calc. Instead provide a Clamp method that returns a new instance with clamped values.  This also prevents extra work if the caller is pre-clamping the values passed to the constructor i.e. move the onus for clamping to the caller.
The clamping in ToSystemDrawingColor is because their doc's specify that the values must by between 0 - 255 i.e. it's undefined behaviour if you passed out-of-bound values.

Typo, won't these be integer division resulting in 0 or 1 ?
this.R = R / 255;  // this.R = R / 255.0f;
float r = systemColor.R / 255;  // float r = systemColor.R / 255.0f;

File: libMBIN/Source/Common/MathHelper.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        /// <summary>
+        /// Clamps a value to be equal to or between the minimum and maximum.
+        /// </summary>
+        /// <param name="value"></param>
+        /// <param name="minimum"></param>
+        /// <param name="maximum"></param>
+        /// <returns></returns>
+        public static float Clampf(float value, float minimum, float maximum)
```
Comment (monkeyman192) [inline libMBIN/Source/Common/MathHelper.cs]: why is this called `Clampf` while all the others are just called `Clamp`? Surely cos it's an overloaded method you just want it to be called `Clamp`?

**Comment by cmkushnir** [comment]
Almost there.  Did you test all the methods ?  I think ToString may be ambiguous due to both being able to be called w/ 0 param, remove the default for FLOAT param i.e. it must be specified in order to use that ToString.

**Comment by cmkushnir** [comment]
Looks good, just looking back over all prior comments, seems missed:
"Instead provide a Clamp method that returns a new instance with clamped values."
e.g. in Colour add a method something like:
public Colour Clamp( float LOWER = 0, float UPPER = 1 ) {
    return new(
        MathHelper.Clamp(R, LOWER, UPPER),
        MathHelper.Clamp(G, LOWER, UPPER),
        MathHelper.Clamp(B, LOWER, UPPER),
        MathHelper.Clamp(A, LOWER, UPPER)
    );
}

**Comment by cmkushnir** [comment]
No more comments from me.

**Comment by monkeyman192** [comment]
Looks good in terms of code. My only issues now are with the docstrings. There are a few instances of `Color` instead of `Colour` (I think it's good to be accurate with this since it could be very confusing otherwise!)
Also, it would be good to elaborate on the summary info for the `ToString` non-overload method to just make it clear what the arguments do (especially the format one so that anyone using it can know what is required of the format. Eg. if it doesn't have 4 slots for variables in it it may raise an error.

**Comment by monkeyman192** [comment]
Ok, looks good to me! If you are happy with the final state of the PR please prefix the title of the PR with `[MRG]` and I'll merge it :)

**Comment by gurrenm3** [comment]
Done

**Pull request merged**